[addons][audioencoder] rework CEncoderFFmpeg to become working again

AlwinEsch opened the pull request:
## Description

At the beginning I only wanted to update it to have the current ffmpeg correct, but I noticed that it no longer worked in the previous version and therefore completely revised it.

In order to avoid the large number of ffmpeg free calls in the event of a failed initialization, I used a try / catch so that it only has to be done in one place.

It's just got really thick with changes :sweat_smile:.

At the moment the commits from https://github.com/xbmc/xbmc/pull/20428 are still in it, as it is based on this, after the other request is in it will be cleaned up here.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
The request now rebased and only includes the ffmpeg cdrip update.

garbear commented on xbmc/cdrip/EncoderFFmpeg.h:
```diff
@@ -33,35 +33,48 @@ class CEncoderFFmpeg : public CEncoder
-  size_t m_ResampledBufferSize{0};
-  AVFrame* m_ResampledFrame{nullptr};
-  bool m_NeedConversion{false};
+  unsigned int m_neededFrames;
```
This variable (and below) looks like it's uninitialized by the default constructor in this class.

AlwinEsch replied:
Will add a 0 set, normally they should not used as done on "Init" and the other using places not called if them fails. but to have set is more correct :+1:.

lrusak commented on xbmc/cdrip/EncoderFFmpeg.h:
```diff
@@ -33,35 +33,48 @@ class CEncoderFFmpeg : public CEncoder
   bool Close() override;
 
 private:
+  struct EncoderException : public std::exception
```
We do have the `Exception` class available here if you don't want to create something new https://github.com/xbmc/xbmc/blob/297b5f844a1db09be4118039e4c29a08e985fb56/xbmc/commons/Exception.h#L44

AlwinEsch replied:
Thanks for the reference to this, had completely forgotten that we have this :+1:.

Just looked over it, think to keep it at this point for now.

Once that the new handler refers purely to his place. "XbmcCommons::Exception" runs a lot through macros and try to avoid such uses if possible.

I think `XbmcCommons::Exception` would be a good new project to revise in the near future and make it conform to the latest C++ standards. Then I think we could integrate it here as well.

The only thing I'm going to do is integrate the string format into the "EncoderException" constructor and avoid the "StringUtils::Format" in the calls.

lrusak reviewed (commented):
Nice cleanup. It's a bit hard to read with the whitespace changes though. Best viewed with whitespace changes disabled. I like the try/catch style and think it's a good alternative to what was used before. It's also a better alternative to my try https://github.com/xbmc/xbmc/pull/16185

AlwinEsch committed:
[addons][audioencoder] rework CEncoderFFmpeg to become working again

At the beginning I only wanted to update it to have the current ffmpeg correct, but I noticed that it no longer worked in the previous version and therefore completely revised it.

In order to avoid the large number of ffmpeg free calls in the event of a failed initialization, I used a try / catch so that it only has to be done in one place.

AlwinEsch committed:
[addons][audioencoder] fix with missing format value

Has badly forgot to add on Log call a "__func__" about format.

AlwinEsch commented on xbmc/cdrip/EncoderFFmpeg.cpp:
```diff
@@ -22,227 +16,233 @@
+{
+  std::string s;
+  template<typename... Args>
+  EncoderException(const std::string& fmt, Args&&... args)
```
Changed on update to allow an string format by exception.

AlwinEsch commented on xbmc/cdrip/EncoderFFmpeg.cpp:
```diff
@@ -22,227 +16,233 @@
+    if (addonId == "audioencoder.kodi.builtin.aac")
+      m_samplesCountMultiply = m_iInSampleRate;
+    else if (addonId == "audioencoder.kodi.builtin.wma")
+      m_samplesCountMultiply = 1000;
```
This parts are added new.
Unfortunately, ffmpeg needs two different ways of setting it. This now sets the multiply at init.

Crazy enough, Kodi also works without it and so I didn't notice it, just had a quick test using VLC and then saw that the length was divided by 48.

AlwinEsch commented:
OK to bring in this evening?
